Make `download*()` methods work in `ContainerGebSpec`

matrei opened the pull request:
Resolves gh-72

matrei committed:
fix: `download*()` methods work

There must be a more elegant way to solve this, but this seems to work.

Resolves gh-72

matrei committed:
fix: Rename `ContainerDownloadSupport` to `ContainerAwareDownloader`

matrei committed:
docs: Improve javadoc for `ContainerAwareDownloader`

matrei committed:
fix: Return correct hostname

matrei committed:
chore: Order constants

matrei committed:
feat: Add `hostName` property for configuring test hostname in `ContainerGebSpec`

Allows specifying `hostName` directly in `ContainerGebSpec` specs, e.g., `String hostName = 'my.test.app'`, to streamline test setup and improve readability.

matrei committed:
refactor: Enhance readability by using state getters

Replaced direct field access with state getters to improve readability and encapsulation. This change makes it clearer how state is accessed and managed within the class.

matrei committed:
refactor: Make `ContainerGebSpec` abstract to prevent direct instantiation

- Converted `ContainerGebSpec` to an abstract class to ensure it is only used as a base class for other specs.
- This change enforces intended usage patterns and prevents accidental instantiation of `ContainerGebSpec`.

matrei committed:
refactor: Rename `ContainerAwareDownloader` to `ContainerAwareDownloadSupport`

Renamed `ContainerAwareDownloader` to `ContainerAwareDownloadSupport` to better reflect that it's a `DownloadSupport` implementation.

jdaugherty reviewed (approved):
I didn’t see another way to do this.  Great Job and thanks for fixing this!

matrei commented:
https://github.com/apache/groovy-geb/issues/204

sbglasius commented on src/testFixtures/groovy/grails/plugin/geb/ContainerGebSpec.groovy:
```diff
@@ -44,38 +47,38 @@ import java.time.Duration
     void setup() {
-        initialize()
-        baseUrl = "$protocol://$hostName:$serverPort"
+        if (notInitialized) initialize()
```
Personally I prefer always to use curly brackets, even if there is just one line to execute in an if... Just a matter of style

sbglasius commented on src/testFixtures/groovy/grails/plugin/geb/ContainerAwareDownloadSupport.groovy:
```diff
@@ -0,0 +1,234 @@
+ * @since 5.0.0
+ */
+@CompileStatic
+trait ContainerAwareDownloadSupport implements DownloadSupport {
```
How about a `@SelfType(ContainerGebSpec)` then it cannot be applied to other `GebSpec`'s

sbglasius reviewed (approved):
You can fix my two suggestions if you want.

matrei committed:
fix: PR feedback

matrei committed:
refactor: Use delegate in `ContainerAwareDownloadSupport`

With this commit `ContainerAwareDownloadSupport` uses an inner class for delegating
to `DefaultDownloadSupport` instead of duplicating all its functionality.

As all `download*()` methods eventually invokes `download(Map options)` it is enough to
override this method, and changing the `base` option in the map to use the proper url.

sbglasius commented:
@matrei Well done with the `@Delegate` instead of dupplicating code.
